### ISSUE TITLE
Partial ordering rules for packs in [temp.deduct.type]

### ISSUE THREAD

**Issue by sdkrystian** [cplusplus/draft#4014]
The wording in [[temp.deduct.type] p10](http://eel.is/c++draft/temp.deduct.type#10) that effectively states "during partial ordering, a function parameter pack is less specialized than `N` matching function parameters" doesn't apply to partial ordering the the context of a function call, since rather than using the entire function type for deduction, only the types of the parameter with arguments in the call are used, meaning that `P` does not have a form that contains `(T)` unless the type of the parameter itself was compounded from a function type.

**Comment by jensmaurer** [comment]
It is correct that the original template argument deduction for function calls is on a per-parameter/argument basis. Using the thus-deduced template parameters, a specialization (of the declaration) is generated which is then used as the input for further overload resolution, which may consider partial ordering as a tie-breaker for the entire candidate (= specialization of the function template declaration). At that point, the entire declaration is in view, and [temp.func.order] p2 says the "function type" is used for partial ordering.

**Comment by sdkrystian** [comment]
@jensmaurer Hm, are you sure? http://eel.is/c++draft/temp.deduct.partial#3 says that the deduction for partial ordering is done on a per parameter/argument basis in the context of a function call

**Comment by jensmaurer** [comment]
Good point.

**Comment by jensmaurer** [comment]
The example given at the end of temp.deduct.type p10 is a non-call context, though. So, where exactly is the editorial problem?

**Comment by sdkrystian** [comment]
The problem itself would arise in the following example:
```cpp
template<typename T>
int f(T); // #1

template<typename... T>
int f(T...); // #2

int a = f(1); // calls #1
```

Our intent here is that `#1` is called, as shown in the example in [[temp.deduct.partial] p8](http://eel.is/c++draft/temp.deduct.partial#8). The partial ordering here is done in the context of a function call, so the deduction is done with respect to the individual parameters and arguments, and consequently the rule that would cause deduction to fail in [[temp.deduct.type] p10](http://eel.is/c++draft/temp.deduct.type#10) (which would make the non-pack parameter at least as specialized as the pack parameter, but not the other way around) doesn't apply.

As for the editorial nature of this issue, that example pretty clearly demonstrates what we want to happen. CWG has already stated what we want to happen in [CWG 1825](http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1825) and [CWG 1395](http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#1395).

**Comment by jensmaurer** [comment]
[temp.deduct.partial] p3 says that we use individual parameters, but then [temp.deduct.partial] p8 has special rules for variadic templates, and look beyond the individual parameter's P/A pair.
The example at the end of p8 is (essentially) the one you quoted above.

**Comment by sdkrystian** [comment]
> but then [temp.deduct.partial] p8 has special rules for variadic templates, and look beyond the individual parameter's P/A pair.

Yup, it effectively says we "look ahead" to match trailing packs. However, it still doesn't use the entire function type, so [temp.deduct.type] p10 wouldn't apply.

**Comment by jensmaurer** [comment]
My assumption is that we don't need/want [temp.deduct.type] p10 to apply for this case, since we already have similar wording in [temp.deduct.partial] p8 for matching variadics.

**Comment by sdkrystian** [comment]
The wording in p8 doesn't cause the deduction to fail (as we intend it to) meaning that it would have no effect on the example I provided.

**Comment by tkoeppe** [comment]
Editorial meeting decision: this is not editorial. Please consider filing a CWG issue.
